### ISSUE TITLE
[FEATURE] Add an additional, even "feather-er" preset

### ISSUE THREAD

**Issue by mikeoliphant** [sdatkinson/neural-amp-modeler#271]
The WaveNet "feather" preset is still too heavy for processors typically used in many audio hardware devices.

It would be very useful to have a fourth preset ("nano" maybe?) designed for these use cases.

I've been experimenting with model architectures, and have settled on a 4x2 (kind of a half-feather) configuration. It shaves 30+% CPU usage compared to "feather" and achieves lower ESR than LSTM models that are comparable in CPU requirements.

I'm happy to submit a PR for this.

**Comment by sdatkinson** [comment]
Sounds good to me.

Only thing I'm not sure I like is making a fourth option. What do you think about changing what "lite" and "feather" mean instead? the current "feather" might be renamed "lite", and the new architecture could be the new "feather"?

**Comment by mikeoliphant** [comment]
> What do you think about changing what "lite" and "feather" mean instead?

That would probably be fine - and maybe preferable. I don't think "lite" is currently used much. My only concern would be that it is a change in behavior - older "feather" models wouldn't be the same as newer "feather" models. Could maybe cause some confusion. That said, I'm not sure there are enough "feather" models currently out there to worry about.

**Comment by sdatkinson** [comment]
Might as well ask the Facebook group--nice to have them around to bounce ideas off of and check our instinct!

**Comment by mikeoliphant** [comment]
> Might as well ask the Facebook group--nice to have them around to bounce ideas off of and check our instinct!

Could help, I guess. The "ESR" isn't always great there, but there definitely is some "signal" to be had.

**Comment by loki42** [comment]
I'd be very keen on this, I'm currently working out what models can run comfortably on Beebo and feather currently uses about 70% of CPU, which is too much to allow a cab / effects as well. If it's a preset in my UI I can say "feather only" that would be great.

**Comment by mikeoliphant** [comment]
@sdatkinson - so far your facebook poll seems to overwhelming favor adding "nano" and leaving the other presets alone (even though hardly anyone uses "lite").

Shall we proceed that way?

At the same time I should probably add a "nano" LSTM config and I may also consider some changes to the existing LSTM presets - I've been experimenting with them a fair bit recently and now have a better idea about what might be useful. My guess is that almost nobody currently knows the LSTM presets are there since we haven't publicized them, so I doubt there will be much if any impact if they change a bit.

**Comment by rominator1983** [comment]
I wonder if this is the right place to ask. I'll go ahead anyway. Could the models maybe be more lightweight if they would be trained with other data?
What I saw is that in all open source projects (SmartPedal, Aida, NAM) there are real guitar/bass recordings for training the models.

If you ever heard what comes out of a quad cortex for capturing you know that this is quite different.

Unfortunately I have hardly the time or experience to test the different approaches against each other. But I know that there are recordings of the training input of some devices around somewhere. So one could make a comparison and decide to create a better/artificial input for training future networks.

**Comment by mikeoliphant** [comment]
> Could the models maybe be more lightweight if they would be trained with other data?

While training data is important, it has no impact on the resulting CPU usage.

> What I saw is that in all open source projects (SmartPedal, Aida, NAM) there are real guitar/bass recordings for training the models. If you ever heard what comes out of a quad cortex for capturing you know that this is quite different.

The official NAM training data is actually more like the what I assume the quad cortex is using - a bunch of frequency sweeps and other signals designed to train a wide variety of responses.

**Comment by loki42** [comment]
Your BossWN-4x2x1.nam is using 54% of a core on Beebo, though I can continue optimising a bit I suspect. The Cortex A53 is less than half the speed of the Pi4 so that tracks pretty well. 

I'm not sure if the numbers are exactly comparable for network type but the LSTM12 is the default Aida configuration and uses 24% of a core, so that also tracks pretty well with your numbers if it's 1x12.

**Comment by KaisKermani** [comment]
I've been testing the provided models [here](https://github.com/mikeoliphant/neural-amp-modeler-lv2/tree/main/models) on the MOD Dwarf (which has a Quad Core ARM A35 CPU at 1.3GHz). These are the CPU consumption numbers I got: 
LSTM 1x16 runs at 34% CPU
LSTM 1x24 runs at 46% CPU
LSTM 2x16 runs at 64% CPU
LSTM 2x8 runs at 38% CPU
LSTM 2x16 runs at 64% CPU
Wavenet 4x2x1 runs at 67%

I would say although all of them can run on the unit, it's much more useful to have models that run at CPU<40%, seeing that most likely it will be used with an IR Cabinet and other effects in the chain.

So out of the proposed architectures, I see that LSTM 1x16 and LSTM 2x8 are the most useful.

Also I'm experimenting on my side with model architectures, and can confirm that if we change the input size of the LSTM layer to 8 (so basically inputting the last 8 samples instead of just the last 1 sample to the network) helps improve the accuracy of the models.

**Comment by mikeoliphant** [comment]
> I would say although all of them can run on the unit, it's much more useful to have models that run at CPU<40%

Depends on the user, I think. Personally, I want the best sounding model that I can get to work reliably with an IR and a few light effects.

I was able to get the Dwarf to run Wavenet 4x2 models in stereo (so two instances on separate cores) through two separate IRs and into stereo delay and reverb. It was pushed hard, but was stable.

**Comment by mikeoliphant** [comment]
> Also I'm experimenting on my side with model architectures, and can confirm that if we change the input size of the LSTM layer to 8 (so basically inputting the last 8 samples instead of just the last 1 sample to the network) helps improve the accuracy of the models.

That's interesting. The NAM LSTM implementation doesn't allow for this currently.

**Comment by KaisKermani** [comment]
My question is, have you experimented with even smaller WaveNet models? maybe it also provides a better CPU-sound quality tradeoff than the LSTM1x16 for example.

**Comment by KaisKermani** [comment]
I was also contemplating the idea of combining both wavenet and LSTM. 
The idea is to have a small dilated conv1d layer in the input, as it best captures the features of input signal over the length of the last 8 or 16 or 32 samples. And then an small LSTM layer (as in 1x8 or 2x4) as it potentially best captures the behavior of the amp in question.

For now this is pretty theoretical, will have to check whether this actually works and provides better results. However since both the convolution and recurrent layers aren't huge i would expect also an improvement in CPU.

**Comment by mikeoliphant** [comment]
> My question is, have you experimented with even smaller WaveNet models? maybe it also provides a better CPU-sound quality tradeoff than the LSTM1x16 for example.

Yes. As I recall, in my testing going below 4x2 didn't give better performance. You are definitely encouraged to experiment, though.

> I was also contemplating the idea of combining both wavenet and LSTM.

I've thought about this a bit as well, but haven't tried anything. It is certainly possible (probable, even) that other potential architectures could provide better CPU/quality tradeoffs.

**Comment by mikeoliphant** [comment]
PR for the new "nano" preset is here: https://github.com/sdatkinson/neural-amp-modeler/pull/279

**Comment by mikeoliphant** [comment]
Resolved via PR https://github.com/sdatkinson/neural-amp-modeler/pull/279

**Comment by sdatkinson** [comment]
> Resolved via PR #279

FYI, writing "Resolves #271" in the PR description makes GitHub automatically close the referenced Issue once the PR is merged 🙂
